### PR TITLE
Support Java 21 main launch protocol

### LINKED ISSUE #36154: Quarkus and Java 21 Instance Main Methods is not working
### Describe the bug

I tried customized QuarkusMain method as shown in https://quarkus.io/guides/lifecycle#the-main-method with Java 21 Instance Main Methods (preview feature).

Code example in https://github.com/rsvoboda/code-with-quarkus/tree/21-main

```java
@QuarkusMain
public class Main {
    void main() {
        System.out.println("==> Running main method");
        Quarkus.run();
    }
}
```

maven-compiler-plugin enabled to use preview features
```xml
    <maven.compiler.release>21</maven.compiler.release>
...
      <plugin>
        <artifactId>maven-compiler-plugin</artifactId>
        <version>${compiler-plugin.version}</version>
        <configuration>
          <compilerArgs>
            <arg>--enable-preview</arg>
          </compilerArgs>
        </configuration>
      </plugin>
```

And I received:
```
Exception in thread "main" java.lang.NoSuchMethodException: org.acme.Main.main([Ljava.lang.String;)
	at java.base/java.lang.Class.getMethod(Class.java:2395)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
```

I checked if I will see `System.out.println("==> Running main method");` when calling the compiled jar just to ensure compilation went well. Message was printed
```
java --enable-preview -cp target/quarkus-app/app/code-with-quarkus-1.0.0-SNAPSHOT.jar org.acme.Main

==> Running main method
Exception in thread "main" java.lang.NoClassDefFoundError: io/quarkus/runtime/Quarkus
	at org.acme.Main.main(Main.java:10)
Caused by: java.lang.ClassNotFoundException: io.quarkus.runtime.Quarkus
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:526)
	... 1 more
```

### Expected behavior

Quarkus boots

### Actual behavior

Quarkus fails to start

### How to Reproduce?

 - Get example code from `21-main` branch - https://github.com/rsvoboda/code-with-quarkus/tree/21-main
 - Run ` mvn clean package -DskipTests && java --enable-preview -jar target/quarkus-app/quarkus-run.jar`

### Output of `uname -a` or `ver`

macOS

### Output of `java -version`

Java 21

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

3.4.1

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by geoand**
This solution is based on transforming the main class so call-sites don't need to change

Closes: #36154
Alternative to: #36158

### CODE REVIEW

**Comment by geoand** [comment]
P.S. we probably want to add some tests for this, but I am very short on time this week. If I do have find some, I'll add them in a follow up PR

**Comment by Ladicek** [comment]
Interesting thing about this is that we'll support the new styles of writing `main` even on older JDKs :-)

(We could also easily support `private` main methods, but I agree we shouldn't.)

**Comment by geoand** [comment]
> Interesting thing about this is that we'll support the new styles of writing main even on older JDKs :-)

Yeah, I thought that was funny as well :). I'll add the `noteworthy-label` for that reason alone :)

**Comment by geoand** [comment]
> P.S. we probably want to add some tests for this, but I am very short on time this week. If I do have find some, I'll add them in a follow up PR

Actually, I see we have similar `QuarkusProdModeTest` type tests, so hopefully I can add some tests in a few minutes

**Comment by Ladicek** [comment]
I just realized there's one or two things this PR doesn't do and at least one of them it should.

One, the `main` method can be `protected`. That's easy to solve, since changing modifiers is just a bunch of bitwise operations on an `int`, so you just do `removeModifiers(PROTECTED)` in addition to `addModifiers(PUBLIC)` and there's no need for additional check.

Two, an instance main can be inherited from a superclass. Not sure if we wanna do anything about that, I guess we can just leave that unsupported. There's an interesting interaction between `static` main declared directly on the class and an instance main inherited from a superclass, so it's probably better to just ignore inherited main than implement it incorrectly.

**Comment by geoand** [comment]
> so you just do removeModifiers(PROTECTED) in addition to addModifiers(PUBLIC) and there's no need for additional check.

good point!

> an instance main can be inherited from a superclass

Ah darn... I do see that https://openjdk.org/jeps/445 mentions a superclass as well... Let me chek what we can do about it..

**Comment by geoand** [comment]
@Ladicek WDYT now? I think the superclass case should be handled now

**Comment by geoand** [comment]
One thing I am not 100% certain of is the proper order of lookups when there are both `main(String[] args)` and `main()` and superclasses involved. 
However I think that's an extreme corner case that we can fix later if need be.

**Comment by Ladicek** [comment]
The JEP [says](https://openjdk.org/jeps/445#Selecting-a-main-method):

> When launching a class, the launch protocol chooses the first of the following methods to invoke:
> - A `static void main(String[] args)` method of non-private access (i.e., `public`, `protected` or package) declared in the launched class,
> - A `static void main()` method of non-private access declared in the launched class,
> - A `void main(String[] args)` instance method of non-private access declared in the launched class or inherited from a superclass, or, finally,
> - A `void main()` instance method of non-private access declared in the launched class or inherited from a superclass.

That seems quite straightforward to me:

1. If the class declares `static void main(String[])`, that's the right one.
2. Otherwise, if the class declares `static void main()`, that's the right one.
3. Otherwise, if the class declares _or inherits_ `void main(String[])`, that's the right one.
4. Otherwise, if the class declares _or inherits_ `void main()`, that's the right one.
5. Otherwise, there's no `main`.

**Comment by geoand** [comment]
Okay, let's try that

**Review by Ladicek — APPROVED** [review]
I would structure the code differently, but it seems OK. The only objection I would have is that there's no test for a combination of multiple valid `main` methods (where the correct one must be picked).

**Comment by geoand** [comment]
Yeah, currentry it does not behave correctly, so I'll need to fix it

**Comment by geoand** [comment]
Should be fixed now (I also added a test to demonstrate it)

**Comment by geoand** [comment]
@Ladicek ok to merge?

File: core/deployment/src/main/java/io/quarkus/deployment/steps/MainClassBuildStep.java [hunk 1]
```diff
@@ -494,4 +504,175 @@ ReflectiveClassBuildItem applicationReflection() {
...
+
+        public MainMethodTransformer(IndexView index) {
+            this.index = index;
+        }
+
+        @Override
+        public ClassVisitor apply(String mainClassName, ClassVisitor outputClassVisitor) {
+            ClassInfo mainClassInfo = index.getClassByName(mainClassName);
+            if (mainClassInfo == null) {
+                throw new IllegalStateException("mainClassName should have a corresponding ClassInfo at this point");
```
Comment (Ladicek) [inline core/deployment/src/main/java/io/quarkus/deployment/steps/MainClassBuildStep.java:523]:
```suggestion
                throw new IllegalStateException(mainClassName + " should have a corresponding ClassInfo at this point");
```

I guess?

Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/steps/MainClassBuildStep.java]: Good catch, fixed

**Comment by Ladicek** [comment]
If I'm reading the code correctly, if the main class defined these 2 methods:

- `void main(String[])`
- `static void main()`

We would prefer the 1st, even though we really should prefer the 1st, as specified in https://docs.oracle.com/javase/specs/jls/se21/preview/specs/unnamed-classes-instance-main-methods-jls.html#jls-12.1.4

**Comment by geoand** [comment]
Ah, that's an interesting edge case indeed...

**Comment by geoand** [comment]
I'm gonna rethink how this is done

**Comment by Ladicek** [comment]
I would just write an extremely stupid code like this:

```
ClassVisitor result = findStaticMainWithStringArgs();
if (result != null) {
    return result;
}
result = findStaticMainWithoutArgs();
if (result != null) {
    return result;
}
result = findInstanceMainWithStringArgs();
if (result != null) {
    return result;
}
result = findInstanceMainWithoutArgs();
if (result != null) {
    return result;
}
throw ...
```

**Comment by geoand** [comment]
That's where I am headed - at least unless I find any other problems :)

**Comment by geoand** [comment]
It was slighly more complicated, but not much. 

I think that anything other than perhaps extreme edge cases should now be covered

**Review by rsvoboda — APPROVED** [review]
Solves https://github.com/quarkusio/quarkus/issues/36154

Thanks @Ladicek and @geoand for the thorough discussion about possible cases + nice impl with attached tests

**Comment by geoand** [comment]
Thanks for reporting and checking!

**Review by Ladicek — APPROVED** [review]
I'm not a fan of renaming those `private` mains to `$originalMain$`, that seems fairly risky and I think it would be better to bail out if we detect a collision. But other than that, it seem fine.

**Comment by geoand** [comment]
I went back and forth on that.
The reason I did it was because I could not find anywhere in the JEP what is supposed to happen in this case.

Bailing out would definitely have been easier from an implementation standpoint 😎

**Pull request merged**